From d3ad42e2778be168cad126ce70dcbf192df9c63c Mon Sep 17 00:00:00 2001
From: Matthias Brugger <mbrugger@suse.com>
Date: Fri, 27 Jan 2017 11:03:54 +0100
Subject: [PATCH] crypto: arm64/crc32 - detect crc32 support in assembler
References: fate#322150
Patch-mainline: never, this got fixed differntly on mainline, but this fixes buils with our toolchain

Older compilers may not be able to detect the crc32 extended cpu type.
Anyway only inline assembler code is used, which gets passed to the
assembler. This patch moves the crc detection to the assembler.

Suggested-by: Alexander Graf <agraf@suse.de>
Signed-off-by: Matthias Brugger <mbrugger@suse.com>
---
 arch/arm64/crypto/Makefile      |    2 --
 arch/arm64/crypto/crc32-arm64.c |    3 +++
 2 files changed, 3 insertions(+), 2 deletions(-)

--- a/arch/arm64/crypto/Makefile
+++ b/arch/arm64/crypto/Makefile
@@ -36,7 +36,5 @@ CFLAGS_aes-glue-ce.o	:= -DUSE_V8_CRYPTO_
 
 obj-$(CONFIG_CRYPTO_CRC32_ARM64) += crc32-arm64.o
 
-CFLAGS_crc32-arm64.o	:= -mcpu=generic+crc
-
 $(obj)/aes-glue-%.o: $(src)/aes-glue.c FORCE
 	$(call if_changed_rule,cc_o_c)
--- a/arch/arm64/crypto/crc32-arm64.c
+++ b/arch/arm64/crypto/crc32-arm64.c
@@ -29,6 +29,9 @@ MODULE_AUTHOR("Yazen Ghannam <yazen.ghan
 MODULE_DESCRIPTION("CRC32 and CRC32C using optional ARMv8 instructions");
 MODULE_LICENSE("GPL v2");
 
+/* Request crc extension capabilities from the assembler */
+asm(".arch_extension crc");
+
 #define CRC32X(crc, value) __asm__("crc32x %w[c], %w[c], %x[v]":[c]"+r"(crc):[v]"r"(value))
 #define CRC32W(crc, value) __asm__("crc32w %w[c], %w[c], %w[v]":[c]"+r"(crc):[v]"r"(value))
 #define CRC32H(crc, value) __asm__("crc32h %w[c], %w[c], %w[v]":[c]"+r"(crc):[v]"r"(value))
